### ISSUE TITLE
Instance metrics config change has no effect

### ISSUE THREAD

**Issue by ksemba** [istio/istio#10209]
**Describe the bug**
I create a custom mixer telemetry adapter / handler and deployed it on GKE according to this guide https://github.com/istio/istio/wiki/Mixer-Out-Of-Process-Adapter-Walkthrough. It worked fine for a few days but now if I modify the dimensions that I want to forward to my adapter it doesn't seem to apply the changes. Whichever dimensions I add or remove from my yaml file, the exact same dimensions get forwarded to my adapter.
Is there a way to restart the mixer to force it to reload the config properly? Using 

> kubectl -n istio-system get instances.config.istio.io i1metric -o yaml

I can verify that my config changes in my yaml also get changed on the istio system config. But still it seems to have no effect on wich dimensions the mixer is forwarding to my adapter. I even deleted my handler, rule and instance but still my adapter continued to receive the same instances with the same old dimensions.

I also tried to follow [https://istio.io/help/ops/telemetry/missing-metrics/#verify-there-are-no-known-configuration-errors](url) 
but setting up the port-forwarding didn't work for me. When I use my browser to go to https://localhost:9093 I only get 

> Handling connection for 9093
E1130 10:22:38.550617   15191 portforward.go:316] error copying from local connection to remote stream: read tcp4 127.0.0.1:9093->127.0.0.1:50672: read: connection reset by peer

**Expected behavior**
When I add a new dimension to my yaml, that new dimension should be reported through my custom metric to my custom adapter. When I remove a dimension it should not be reported anymore.

**Steps to reproduce the bug**
I tried adding and removing dimension from my yaml through adding or removing the hashtags in-front of the corresponding dimenion config line:

# handler for adapter mygrpcadapter
apiVersion: "config.istio.io/v1alpha2"
kind: handler
metadata:
 name: analytics-cloud
 namespace: istio-system
spec:
 adapter: analytics-cloud
 connection:
         address: "analytics-cloud.istio-system:8090"
 params:
   file_path: "out.txt"

# instance for template metric
apiVersion: "config.istio.io/v1alpha2"
kind: instance
metadata:
 name: i1metric
 namespace: istio-system
spec:
 template: metric
 params:
   value: request.size | 0
   dimensions:
     reporter: conditional((context.reporter.kind | "inbound") == "outbound", "client", "server")
     #request_headers: request.headers
     #request_path: request.path | "unknown"
     #request_host: request.host | "unknown"
     #request_referer: request.referer | "unknown"
     #request_size: request.size | 0
     #request_total_size: request.total_size | 0
     #request_time: request.time
     #request_useragent: request.useragent | "unknown"
     #response_headers: response.headers
     #response_size: response.size | 0
     #response_total_size: response.total_size | 0
     #response_time: response.time
     #response_duration: response.duration
     #response_code: response.code | 200
     #source_name: source.name | "unknown"
     #source_workload_name: source.workload.name | "unknown"
     #source_service: source.service | "unknown"
     #source_version: source.labels["version"] | "unknown"
     #source_ip: source.ip #| "unknown"
     #source_port: source.port | 0
     #destination_name: destination.name | "unknown"
     #destination_workload_name: destination.workload.name | "unknown"
     #destination_service: destination.service | "unknown"
     #destination_version: destination.labels["version"] | "unknown"
     #destination_ip: destination.ip #| "unknown"
     #destination_port: destination.port | 0
     #connection_event: connection.event | "unknown"
     #connection_received_bytes: connection.received.bytes | 0
     #connection_received_bytes_total: connection.received.bytes_total | 0
     #connection_sent_bytes: connection.sent.bytes | 0
     #connection_sent_bytes_total: connection.sent.bytes_total | 0
     #connection_duration: connection.duration
     #context_protocol: context.protocol
     #context_time: context.time

---
# rule to dispatch to handler h1
apiVersion: "config.istio.io/v1alpha2"
kind: rule
metadata:
 name: rule-analytics-cloud
 namespace: istio-system
spec:
 actions:
 - handler: analytics-cloud.istio-system
   instances:
   - i1metric
---

**Version**
kubectl client 1.12, server: 1.9+

**Installation**
downloaded it from git and build it according to the dev guide for custom adapters

**Environment**
Google Kubernetes Engine

**Comment by bianpengyuan** [comment]
It is surprising that with rule, handler and instance deleted your adapter is still getting report calls. Would you mind to kubectl get -n istio-system {rule, handler, instance} to make sure there is no garbage config left?

**Comment by douglas-reid** [comment]
@ksemba my instinct is that there is a configuration error in your updated config. If you are not able to get port-forwarding to work for mixer, can you try querying for the metrics via prometheus? We are interested in values for metrics like `mixer_config_rule_config_errors_total` and `mixer_config_rule_config_errors_total` and `mixer_config_rule_config_errors_total` and `mixer_config_handler_validation_error_total` (though you might just want to look around at all of the `mixer_config_*` values).

Also, do you see anything in the logs for `istio-telemetry`?

I'm not clear on the hashtags bit. Do you mean, you've tried commenting out lines to no luck?

I agree with @bianpengyuan that if you deleted the rule for your instance and were still receiving instances, something is likely up -- which may signify config failures.

**Comment by ksemba** [comment]
@bianpengyuan if I try to delete my custom istio config (that yaml deletes the handler, rule and metric instance but not my adapter) I get the following output:

> kubectl delete -f kube_istio_deployment.yaml
handler.config.istio.io "analytics-cloud" deleted
instance.config.istio.io "i1metric" deleted
rule.config.istio.io "rule-analytics-cloud" deleted

So it seems that all three objects are deleted. But the adapter seems to still live:

> kubectl get adapters.config.istio.io --all-namespaces
NAMESPACE      NAME                      AGE
istio-system   analytics-cloud   12d
istio-system   mygrpcadapter             13d

--> my rule is gone

> kubectl get rules --all-namespaces
NAMESPACE      NAME                     AGE
istio-system   kubeattrgenrulerule      40d
istio-system   promhttp                 40d
istio-system   promtcp                  40d
istio-system   stdio                    40d
istio-system   stdiotcp                 40d
istio-system   tcpkubeattrgenrulerule   40d

--> my instance is also gone:

> kubectl get instances.config.istio.io --all-namespaces
No resources found.

@douglas-reid I might need some help on you to query those metrics from prometheus. Since the prometheus service has no external IP and I cannot do port-fowarding I tried to use a busybox to run wget queries against its internal service IP. But:

> kubectl exec busybox -- wget 'http://10.59.243.98:9090/api/v1/query?query=up&time=2018-12-01T20:10:51.781Z'
Defaulting container name to busybox.
Use 'kubectl describe pod/busybox -n default' to see all of the containers in this pod.
Connecting to 10.59.243.98:9090 (10.59.243.98:9090)
wget: server returned error: HTTP/1.1 503 Service Unavailable
command terminated with exit code 1

The example query I used was taken from here: https://prometheus.io/docs/prometheus/latest/querying/api/

The istio telemetry logs (kubectl -n istio-system logs $(kubectl -n istio-system get pods -lapp=telemetry -o jsonpath='{.items[0].metadata.name}') -c mixer) don't show that much. I tried deleting my rule/instance/handler and re-create them and looked at the telemetry logs but saw nothing interesting. When there is telemetry data to deliver to my adapter, it shows that data. the only somewhat interesting logs I found were these:

> 2018-12-02T18:04:54.610529Z     info    adapters        Cache sync successful.  {"adapter": "handler.kubernetesenv.istio-system"}
2018-12-02T18:04:54.614374Z     info    Cleaning up handler table, with config ID:56
2018-12-02T18:04:54.614677Z     error   adapters        adapter did not close all the scheduled daemons {"adapter": "handler.kubernetesenv.istio-system"}
2018-12-02T18:04:54.614985Z     info    transport: loopyWriter.run returning. Err: connection error: desc = "transport is closing"
2018-12-02T18:04:55.614945Z     info    Publishing 8 events
2018-12-02T18:04:55.627997Z     info    Built new config.Snapshot: id='58'
2018-12-02T18:04:55.628476Z     info    adapters        getting kubeconfig from: ""     {"adapter": "handler.kubernetesenv.istio-system"}
2018-12-02T18:04:55.628669Z     warn    Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2018-12-02T18:04:55.630127Z     info    adapters        Waiting for kubernetes cache sync...    {"adapter": "handler.kubernetesenv.istio-system"}
GC forced

But they are from yesterday and I didn't modify my setup yesterday...

The commented lines in my kube_istio_deployment.yaml's dimensions section are just to test to de-/activate various dimensions. I'm trying to push more and more data to my adapter in order to parse and process it...

**Comment by ksemba** [comment]
I tried to use the original YAML file that I created when I setup my istio dev environment (copied from the istio documentation) and with that file everything started working again. I used three different file editors to edit that file over time (vi and Visual Studio under Linux and Notepad++ on Windows) and maybe that messed up the file (-format) somehow. Notepad++ shows the file format as Linux (LF) and UTF-8 and I couldn't spot any Windows-specific characters but that is the best explanation that I have.

Since it is now working I tried expanding the list of dimensions that I send to my adapter. Most of them work but as soon as I add one of these, the adapter only receives the "reporter" dimension but none of the others:

> source_port: source.port | 0
connection_duration: connection.duration
context_time: context.time

Not sure what's wrong with these dimensions ... 

I also added these two to my YAML but they don't get reported (but also don't break the reporting of all the other, working ones): 

> request_headers, response_headers

**Comment by ksemba** [comment]
@bianpengyuan @douglas-reid @kyessenov I've been playing with the instances that I forward to my adapter by copying the rule that is used within the prometheus adapter into my YAML file:

> apiVersion: "config.istio.io/v1alpha2"
kind: rule
metadata:
 name: analytics-cloud-http
 namespace: istio-system
spec:
 match: context.protocol == "http" || context.protocol == "grpc"
 actions:
 - handler: analytics-cloud.istio-system
   instances:
   - requestcount.metric
   - requestduration.metric
   - requestsize.metric
   - responsesize.metric

That seemed to have messed up my environment. Now my adapter only receives instances that have the request.size as value and the reporter as the only dimension. Even after applying the YAML that I used before (which defines lots of dimensions) I only get one dimension. 

> received request {[&InstanceMsg{Value:&istio_policy_v1beta11.Value{Value:&Value_Int64Value{Int64Value:0,},},Dimensions:map[string]*istio_policy_v1beta11.Value{reporter: &Value{Value:&Value_StringValue{StringValue:client,},},},MonitoredResourceType:,MonitoredResourceDimensions:map[string]*istio_policy_v1beta11.Value{},Name:i1metric.instance.istio-system,}] &Any{TypeUrl:type.googleapis.com/adapter.analytics.cloud.config.Params,Value:[10 7 111 117 116 46 116 120 116],XXX_unrecognized:[],} 11922098934566016383}

I then tried to delete the rule, the instance and the handler but my adapter log 

> kubectl logs -f $(kubectl get pod -l app=analytics-cloud -o jsonpath="{.items[0].metadata.name}" -n istio-system) -c analytics-cloud-container -n istio-system

still shows new instance data coming in when I refresh the bookinfo web page! 
Even though the rule that tells mixer to deliver instances to my handler doesn't exist anymore:

> kubectl get rules --all-namespaces                                                                                                NAMESPACE      NAME                     AGE
istio-system   kubeattrgenrulerule      41d
istio-system   promhttp                 41d
istio-system   promtcp                  41d
istio-system   stdio                    41d
istio-system   stdiotcp                 41d
istio-system   tcpkubeattrgenrulerule   41d

How is this even possible? Cached config that still lives on in mixer?
